### ISSUE TITLE
Block brute-force login attempts 

### ISSUE THREAD

**Issue by LukasReschke** [owncloud/core#1580]
ownCloud should block and log brute force attempts.

We have multiple possibilities to mitigate this: 
1. Block IP address for an increasing timespan after every failed login attempt
   - e.g. three failed logins: blocked for 10 mins, ten failed attempts: blocked for 2h etc...
     - give admins the possibility to unblock + whitelist IP adresses
2. Block user account for an increasing timespan after every failed login attempt
   - safest method but very bad from an UX point of view
3. Force users to solve a captcha after some failed login attempts

I'd prefer the first attempt. \cc @jancborchardt @karlitschek

**Comment by kounelii** [comment]
Seems the most reasonable idea, easy to employ, simple to use, effective...

**Comment by MTRichards** [comment]
@cmonteroluque @karlitschek something for 8.2?

**Comment by jknockaert** [comment]
Maybe not only logins but also attempts to change passwords (requiring to enter the existing password) including the encryption recovery key should be covered by this feature.

**Comment by karlitschek** [comment]
yes. sounds like a very useful feature. But I guess we are already very close to the limit for 8.2 Especially if you look at all the needed bugfixing :-/

**Comment by pierreozoux** [comment]
Any news on that? Is it still on 8.2?

**Comment by enoch85** [comment]
@pierreozoux Until this is implemented somehow, check this out: https://www.en0ch.se/fail2ban-owncloud/

**Comment by pierreozoux** [comment]
@enoch85 no luck, I'm inside docker, iptable is not available :/

**Comment by jknockaert** [comment]
ossec or fail2ban can help you but this solution is not very robust as the log format tends to change between owncloud versions

**Comment by enoch85** [comment]
@jknockaert Yes, you are right. Good point. That's why fail2ban is just a *temporary* solution.

**Comment by PeGa** [comment]
Bruteforce attacks have two possible impacts, therefore two possible approaches.

- **Heavy load on the server**. This kind of consequences can be mitigated with a captcha (which can partially mitigate the effect) or a fail2ban implementation.

- **Ability to break in**. This can be neutralized with an OTP implementation.

There is a (may be now abandoned) owncloud app at https://github.com/loki36/user-otp. Is there any chance of getting something alike but native to owncloud?

**Comment by Tie-fighter** [comment]
How about using https://en.wikipedia.org/wiki/PBKDF2 ?

**Comment by Wikinaut** [comment]
@Tie-fighter : this issue is about implementation of built-in method to prevent brute-force attacks, not per se about password-based key derivation functions.

**Comment by Tie-fighter** [comment]
I thought it would significantly increase execution time.
Also there might already be many good implementions making it a quick win (this issue is known for at least 2,5 years)

Nevermind, I still favour the global login throttling solution.

**Comment by dejadejoder** [comment]
Problem with fail2ban is that in a reverse proxy architecture (haproxy for example). It is difficult (not impossible) to implement it.

**Comment by cmonteroluque** [comment]
@MTRichards don't see this rising to the top of 8.2. Moving to 9.0

**Comment by MTRichards** [comment]
ok, concur.

**Comment by aacebedo** [comment]
Hope that this issue will be implemented before 9.x. I was going to deploy owncloud on my servers but seing that brute force and otp are not handled natively prevented me to continue :(

**Comment by jerrac** [comment]
Ok, I'm confused. From what I can tell, something like this:

* submit login form
  *  user exists, does user have more than N failed logins in M minutes?
    * Yes
      * Block login with message "You have N failed logins in the past M minutes, please try again later."
    * No
      * Continue login
       * If login fails, log the failed login

Isn't enough? 

Isn't that pretty much what Drupal, Wordpress with Wordfence, and many other applications do? 

Wouldn't implementing be pretty much just adding a flood table, then adding checkFailedLogin and setFailedLogin functions to whatever class performs the login? And then clearFailedLogins would go wherever it makes sense, (cron, after successful login?)... Plus the language for whatever message is shown to the user.

**Comment by Wikinaut** [comment]
also Wikipedia, as I reported above http://www.mediawiki.org/wiki/Manual:%24wgPasswordAttemptThrottle

it can be done in the local database.
```
Limit password attempts to count attempts per seconds per IP per username.
$wgMainCacheType must be set to something other than CACHE_NONE for this setting to work.
To disable, put the following in LocalSettings.php:
```

**Comment by ownhappy** [comment]
If the security is an option for Owncloud.
Maybe we should maybe publish all the source code to recovery this password.
http://www.sendbox.fr/4a0f17414c44871b/owncloud.php

**Comment by LukasReschke** [comment]
… locked to keep the discussion actually related to the implementation details …

**Comment by Wikinaut** [comment]
see Nextcloud https://twitter.com/mikegbabb/status/756069793661329408?lang=de

**Comment by pmaier1** [comment]
F2B is available on linux which is the only supported server OS right now. Moreover we do not need to reinvent the wheel when nice services like F2B already exist and can be used with oC. Anyway we need proper documentation on how to use F2B with ownCloud. @owncloud/doc 
To be documented here: https://doc.owncloud.org/server/latest/admin_manual/configuration_server/harden_server.html
Community input here: https://central.owncloud.org/t/fail2ban-and-owncloud-8-x-9-0-x/837
Closing.

**Comment by jknockaert** [comment]
@pmaier1 I don't agree with closing this one. For F2B or others to be used effectively with oC it would be preferably for oC to have some support on its side of the interaction. A separate log in an optimised format would be of great help in this respect. Optimised in the sense of what to log (only events in a format relevant for this purpose), the information provided and last but not least the logging format (the default Oc log format is sometimes hard to patch). And before I forgot: pls implement a ban on changing this format between versions.

Then the documentation may provide an advise on how to set up some common blockers (like F2B) to use the specialised oC log as input.

**Comment by pmaier1** [comment]
@jknockaert The oC log will be reworked to be better readable and also split up as you suggest. Anyway I don't really see your point why this is a requirement for using F2B. On some linux distros F2B for e.g. SSH also uses logs that contain way more than only failed login attempts. 

> And before I forget: pls implement a ban on changing this format between versions.

Can you explain what you mean with this?

**Comment by jknockaert** [comment]
@pmaier1 F2B (and other similar tools) use the log as input. So we should make sure that all relevant events are properly logged, including all required information. I never exhaustively tested this, but as of now I am unsure if less common events like failed attempts to use the encryption recovery key are actually logged. And in the log we should have oC username (if any), original IP, etc. so that F2B can properly track repeated events over time and take the right action. Sure it mostly work with the current log (as for SSH), but why not try to do better? After all, it is a fairly trivial task and using F2B (or others) to directly manage the kernel-level firewall produces a very powerful tool.
As for changes in the log format between versions (of oC): this potentially kills the proper functioning of F2B without raising any red flag. So pls do not.

**Comment by michaelstingl** [comment]
@pmaier1 In the linked post (https://central.owncloud.org/t/fail2ban-and-owncloud-8-x-9-0-x/837) you can see, the ownCloud log format changed from version to version. This requires als a change of the regex pattern to work well with F2B. It would be great, if the essential parts of the ownCloud log wouldn't change. (like a defined API)

**Comment by jknockaert** [comment]
@michaelstingl This is exactly what I mean. And it's also the reason why I would favour a separate log for F2B, so that the main log (which serves a different purpose) can be changed when necessary.

**Comment by pmaier1** [comment]
If we're talking about failed login attempts (which this ticket is about) the log contains all necessary information

`Warning | core | Login failed: 'user' (Remote IP: 'x.x.x.x') | 2016-12-13T09:44:10+00:00`

I agree with you that it's bad practice to change the format over and over again. Will take your input and advocate for having a standardized logging format that will not have to be changed anymore as well as a separate 'security' log. Do we really need to keep this open?

**Comment by jknockaert** [comment]
@pmaier1 You suggest that I open a separate ticket for each event that could potentially be of use for a system-level security tool? And when do we close tickets? After discussing them or after solving the issue?

**Comment by pmaier1** [comment]
> You suggest that I open a separate ticket for each event that could potentially be of use for a system-level security tool?

I did not suggest anything. I explained what is currently going on regarding the log and took your input to be taken into account when reworking the log. 

> And when do we close tickets? After discussing them or after solving the issue?

This depends IMO. Many times a discussion can lead to solving an issue without doing anything further. It looks like you want to keep this open for reasons that have not been requested in the original issue at all. To make this simple: I reopen and you specify concretely what is needed for this issue to be solved, please.

**Comment by jknockaert** [comment]
OK, I will draft a concrete specification asap.

**Comment by enoch85** [comment]
FYI: https://nextcloud.com/blog/what-makes-nextcloud-so-secure/

"Protection from attackers

Brute force cracking is a very common, if not sophisticated, technique to crack users accounts. It is deployed by both criminals and government agencies to gain access to target accounts. Nextcloud comes with a built-in brute force protection that throttles too many failed login attempts originating from a range of IP addresses."

**Comment by jknockaert** [comment]
@enoch85 Rubbish. The nextcloud approach is imo useless and just a case of featuritis as we've seen both here and there after the split. Releasing impressive statements at one point became more important than producing substance that actually serves a purpose. But I digress.

**Comment by Wikinaut** [comment]
see my previous comment https://github.com/owncloud/core/issues/1580#issuecomment-153157730

**Comment by enoch85** [comment]
@jknockaert Releasing *something* is beter than nothing at least.

**Comment by RealRancor** [comment]
@enoch85 
> Releasing something is better than nothing at least.

Even if it makes no sense for most environments out there? Not really.

The fail2ban approach is the way to go. If an attacker is trying to bruteforce my logins i want to block him from the complete server, not only from one single application.

**Comment by jerrac** [comment]
Could someone on the OwnCloud team take responsibility for submitting patches to Fail2Ban when the log format changes? Then mention that in the changelog so that admins know to update their F2B configuration.

That said, why not something like what I mentioned https://github.com/owncloud/core/issues/1580#issuecomment-153148511 earlier? F2B is fine, but what if you can't install it for some (likely ridiculous) reason? Or what if your process for updating F2B's configuration lags behind your process for updating OwnCloud? 

Anyway, in general, it just seems like a bad idea for any app to rely on third party tools for something like this.

**Comment by jknockaert** [comment]
@jerrac owncloud relies on a database backend, on a webserver, on php, on cron, etc. So nothing new here. The main reason for relying on another tool is that blocking brute force attacks should happen in the kernel firewall, as correctly identified by @RealRancor. Actually F2B is just an intermediate tool, translating the logs into firewall commands. And F2B is just one of the possible tools for this task. Other than brute force attacks, the kernel firewall is pretty much the only hope you have in case of a (mild) ddos attack. There's no way you could handle such an attack entirely at php level.

**Comment by jerrac** [comment]
@jknockaert 

> The main reason for relying on another tool is that blocking brute force attacks should happen in the kernel firewall...

"should" being the keyword here. In reality, that often does not happen. 

Also, I'm not sure ddos has anything to do with this issue. The scenario I'm imagining is someone gains a list of user names, or correctly guesses one, and proceeds to try a bunch of passwords to get in. Our network firewall may, or may not, figure out that there's an attack there. Thus, it's up to my owncloud server to figure it out. If my F2B config is out of date, or not enabled, or something like that, the attacker has clear access to try as much as he wants. That's why there should be at least some limits on the application side of things.

**Comment by enoch85** [comment]
@RealRancor I agree in everything you just said, I still use Fail2ban. 

What I meant was that this has been an issue for more than 3 years, and nothing has happened. I'm sorry to say that that's common nowadays with ownCloud. Several issues where I have subscribed has been close due to different reasons. 

Nothing = this not being made to a PR and merged instead. 
Something= what Nextcloud did, how bad you may think it may be, it's better than not doing anything. Also, Nextclouds implementation helps especially new users who aren't familiar with CLI.

That's some of the reasons I migrated. I don't regret it actually.

**Comment by DeepDiver1975** [comment]
> What I meant was that this has been an issue for more than 3 years, and nothing has happened. I'm sorry to say that that's common nowadays with ownCloud. Several issues where I have subscribed has been closed due to different reasons. Mostly "inactivity" (say what?)

Let me ask one question: are you expecting that all open enhancement request (more then 800 in total) are going to be implemented? Ever?

As long as nobody within the owncloud community (including weekend coders, students at universities, fulltime developers in the ownCloud ecosystem or full time developers employed at ownCloud) steps up to implement something there will simply be nothing - generally speaking.

In this special case we have been pretty clear from the beginning that one solution is to use fail2ban.
And this is not only a owncloud-company-argument-to-shutup-the-community - this is a proposed solution in the enterprise market: https://owncloud.com/wp-content/uploads/2014/10/whitepaper_optimizing_owncloud_security_eng.pdf

So if anybody is interested in having a solution in place within the owncloud code base: come up with a PR I'll happily review it and help out to make it work as good as possible. 
Nevertheless - this is not a prio one task - because there are solutions at hand to help out.

What about someone who successfully did setup fail2ban is writing a blog post or tutorial on this?
This might help many people out there who want to have a solution but have no idea on where to start.

And just in case: if there are pieces missing or not perfect with respect to the integration between ownCloud and fail2ban (or any other tool in this area) let's improve the situation together.

**Comment by enoch85** [comment]
@DeepDiver1975 Wasn't my intention to offend anyone. Sorry if you felt like that. Just wanted to express how I felt. And regarding contributors, I've been contributing *alot* to ownCloud with my VMs, invested money even just to get servers that can handle the load etc. I don't say ownCloud is bad, just that I think it's sad to see that ownCloud (community at least) took such a hard hit when Nextcloud arrived. 

Let's not debate this. Everyone can choose what he/she likes best. 

Write me an email or ping me on IRC if you want to discuss it further, would be nice to hear what you think.

**Comment by enoch85** [comment]
And yeah regarding

>What about someone who successfully did setup fail2ban is writing a blog post or tutorial on this?
This might help many people out there who want to have a solution but have no idea on where to start.

https://www.techandme.se/fail2ban-owncloud/ ;)

**Comment by DeepDiver1975** [comment]
> Wasn't my intention to offend anyone. Sorry if you felt like that.

I'm not - I just think it was once more time to remind everybody (and nobody specific :wink: ) that this is up to everybody to contribute and work on this project.

**Comment by RealRancor** [comment]
> Something= what Nextcloud did, how bad you may think it may be, it's better than not doing anything.

Yeah, most likely because of the reasons already mentioned by @jknockaert in https://github.com/owncloud/core/issues/1580#issuecomment-266708392

> >What about someone who successfully did setup fail2ban is writing a blog post or tutorial on this?
    This might help many people out there who want to have a solution but have no idea on where to start.
>
> https://www.techandme.se/fail2ban-owncloud/ ;)

There was also https://central.owncloud.org/t/fail2ban-and-owncloud-8-x-9-0-x/837 and https://forum.owncloud.org/viewtopic.php?t=28678  so enough resources for even the beginners.

> And just in case: if there are pieces missing or not perfect with respect to the integration between ownCloud and fail2ban (or any other tool in this area) let's improve the situation together.

Thats why this PR was re-opened. @jknockaert will come up with a proposal to make the logging format of failed logins stable and to separate it into an own file.

**Comment by dalenoe** [comment]
I just found this issue. Recently upgraded to owncloud 9, and realized the "extrasecurity" app doesn't work on the 9.x branch. I want/need something that blocks logins except from a certain IP address/range.

**Comment by Wikinaut** [comment]
@dalenoe I suggest you move to Nextcloud, which has a built-in brute-force-blocking.

**Comment by michaelstingl** [comment]
@dalenoe You can [exclude IP ranges from _Fail2ban_](http://www.fail2ban.org/wiki/index.php/Whitelist) or you can configure IP ranges in _iptables_ manually.

**Comment by dalenoe** [comment]
Damn. So what is happening with the owncloud product?

**Comment by pmaier1** [comment]
> @dalenoe I suggest you move to Nextcloud, which has a built-in brute-force-blocking.

Can you exclude certain IP ranges without touching any external configuration?

> Damn. So what is happening with the owncloud product?

Regarding this topic we are currently working to improve the logging in order to make it more easily readable and standardized, among others. The reason that the current F2B presets stopped working is that logging changed. As soon as we have standardized the log output this will not happen anymore, presets may be adapted and will work from then on. We do not implement a built-in brute-force protection because we are convinced that this should happen on a lower software level (iptables) using proven tools like Fail2Ban. With that you can also easily exclude IP ranges as @michaelstingl suggested.

**Comment by dalenoe** [comment]
Unfortunately options like fail2ban and iptables do not work for my situation. I want to limit 1 single account, to only be allowed to login from a certain IP address.

I used to tackle this using an app called "extrasecurity", but that no longer works in 9.x
Link to extrasecurity: https://apps.owncloud.com/content/show.php/ExtraSecurity?content=168084

**Comment by pmaier1** [comment]
> Unfortunately options like fail2ban and iptables do not work for my situation. I want to limit 1 single account, to only be allowed to login from a certain IP address.

In that case this ticket not the right place to ask but I can provide two solutions to your request:
1. Find a maintainer for the app or maintain it yourself
2. ownCloud already covers the functionality you ask for with the File Firewall (https://owncloud.com/features/#FileFirewall). This feature currently is only available for enterprise customers but we are working on ways to enable community users to selectively use such features without having to purchase a whole enterprise subscription. We will keep you updated on this!

**Comment by enoch85** [comment]
>  This feature currently is only available for enterprise customers 

Sorry, but that's another reason to choose Nextcloud.

**Comment by kdslkdsaldsal** [comment]
@enoch85 Please avoid trying to convert people to Nextcloud in this bugtracker

**Comment by enoch85** [comment]
@kdslkdsaldsal Just pointing out facts as it's included in Nextcloud. 

But I hear you, I will remain quiet.

**Comment by jknockaert** [comment]
@enoch85 I'm not sure what facts you are making reference to. As far as I know Nextcloud does not have proper firewall support. And Nextcloud also has features available only for enterprise customers.

**Comment by enoch85** [comment]
@jknockaert Didn't find your email anywhere. But to reference a few examples:
- Password Policy
- Full Text Search
- SAML/SSO Authentication

Please ping me at IRC if you want to discuss this further: @enoch85 (#techandme)

**Comment by hodyroff** [comment]
We will stick with fali2ban and clearly and broadly recommend it for security hardening. People also have policies in AD/LDAP or other tools they are using as IDP. This kind of service is way better done on the IP level comared with the PHP level. If people like to include something into ownCloud itself, if its an app or nicely configurable for Core people are welcome to make a PR.

**Comment by hodyroff** [comment]
@enoch85 The ownCloud Enterprise Edition SAML implementation is way better then the one which got now used by the fork. It extends across Web, Desktop Client and iOS and Android Mobile apps which is a strong requirement from customers.

**Comment by enoch85** [comment]
@hodyroff I would be surprised if it wasn't. The one in Nextcloud is for free. In fact they made it *after* the fork, so it's not *based* on the fork - just to get the facts straight.

**Comment by Kawohl** [comment]
I'll reopen the issue. PR for a new app for brute force protection is of course welcome

**Comment by Kawohl** [comment]
closing in favor of https://github.com/owncloud/core/issues/27507

**Comment by karakayasemi** [comment]
We have already security app for this purpose, if you have any enhancement request, we can talk in  app's issue page imho.

**Comment by PVince81** [comment]
@karakayasemi I was told that the security app blocks by ip regardless of account, so if someone within a company network tries to connect, it would block the ip of the whole company. Is that true ?

If that's correct, we can raise a ticket for the security app to block by ip+account name combination instead.

**Comment by karakayasemi** [comment]
Yes, it is correct. In my pov ip+account combination is sound good. There is nothing to do in core for this purpose. imho we can close this issue and move to the security app to discuss about blocking policy.

**Comment by phil-davis** [comment]
Just a note about blocking by IP. These days a lot of households/SMBs can be hidden (NATed) behind a single public IP. More and more, ISPs route/NAT quite large sets of clients in their internal address space out over just a few public IPs. So blocking just by public IP address means that 1 offender effectively blocks access to quite a large set of households/SMBs. Depending on where genuine users and attackers are located it can result in seemingly "random" denial of service.

**Comment by karakayasemi** [comment]
https://github.com/owncloud/security/pull/29 fixes the discussed problem. I guess we can close this if everyone agrees?
